cmd/kubeadm/app/preflight/checks.go

```diff
@@ -957,11 +957,14 @@ func RunJoinNodeChecks(execer utilsexec.Interface, cfg *kubeadmapi.NodeConfigura
 // addCommonChecks is a helper function to deplicate checks that are common between both the
 // kubeadm init and join commands
 func addCommonChecks(execer utilsexec.Interface, cfg kubeadmapi.CommonConfiguration, checks []Checker) []Checker {
-	// Check whether or not the CRI socket defined is the default
-	if cfg.GetCRISocket() != kubeadmdefaults.DefaultCRISocket {
-		checks = append(checks, CRICheck{socket: cfg.GetCRISocket(), exec: execer})
+	containerRuntime, err := kubeadmutil.NewContainerRuntime(execer, cfg.GetCRISocket())
+	if err != nil {
+		glog.Errorf("[preflight] %v", err)
 	} else {
-		checks = append(checks, ServiceCheck{Service: "docker", CheckIfActive: true})
+		checks = append(checks, ContainerRuntimeCheck{runtime: containerRuntime})
+		if containerRuntime.IsDocker() {
+			checks = append(checks, ServiceCheck{Service: "docker", CheckIfActive: true})
+		}
 	}
 
```

Comment:
Look at the SystemVerification check; it will add a docker check if we are using docker, but the way it determines that is `if sysver.CRISocket == kubeadmdefaults.DefaultCRISocket {` which should be cleaned up in this PR

Reply:
@chuckha hmm, still don't get it. Are you talking about this change?
```
@@ -532,8 +525,11 @@ func (sysver SystemVerificationCheck) Check() (warnings, errors []error) {
 	var validators = []system.Validator{
 		&system.KernelValidator{Reporter: reporter}}
 
-	// run the docker validator only with dockershim
-	if sysver.CRISocket == kubeadmapiv1alpha3.DefaultCRISocket {
+	// run the docker validator only with docker runtime
+	containerRuntime, err := utilruntime.NewContainerRuntime(utilsexec.New(), sysver.CRISocket)
+	if err != nil {
+		errs = append(errs, err)
+	} else if containerRuntime.IsDocker() {
 		// https://github.com/kubernetes/kubeadm/issues/533
 		validators = append(validators, &system.DockerValidator{Reporter: reporter})
 	}
```

What's wrong with it?

Reply:
@luxas ah, got it at last! will do. thank you for the suggestion!